dill/_utils.py

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/env python
+#
+# Author: Leonardo Gama (@leogama)
+# Copyright (c) 2022 The Uncertainty Quantification Foundation.
+# License: 3-clause BSD.  The full license text is available at:
+#  - https://github.com/uqfoundation/dill/blob/master/LICENSE
+"""auxiliary internal classes used in multiple submodules, set here to avoid import recursion"""
+
+__all__ = ['AttrDict', 'ExcludeRules', 'Filter', 'RuleType']
+
+import logging
+logger = logging.getLogger('dill._utils')
+
+class AttrDict(dict):
+    """syntactic sugar for accessing dictionary items"""
+    def _check_attr(self, name):
+        try:
+            super().__getattribute__(name)
+        except AttributeError:
+            pass
+        else:
+            raise AttributeError("'AttrDict' object attribute %r is read-only" % name)
+    def __getattr__(self, key):
+        # This is called only if dict.__getattribute__(key) fails.
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError("'AttrDict' object has no attribute %r" % key)
+    def __setattr__(self, key, value):
+        self._check_attr(key)
+        self[key] = value
+    def __delattr__(self, key):
+        self._check_attr(key)
+        del self[key]
+    def __reduce__(self):
+        return type(self), (dict(self),)
+
+
+### Namespace filtering
+import re
+from dataclasses import InitVar, dataclass, field, fields
+from collections import abc, namedtuple
+from enum import Enum
+from functools import partialmethod
+from itertools import filterfalse
+from typing import Callable, Iterable, Pattern, Set, Tuple, Union
+
+RuleType = Enum('RuleType', 'EXCLUDE INCLUDE', module=__name__)
+NamedObj = namedtuple('NamedObj', 'name value', module=__name__)
+
+Filter = Union[str, Pattern[str], int, type, Callable]
+Rule = Tuple[RuleType, Union[Filter, Iterable[Filter]]]
+
+def isiterable(arg):
+    return isinstance(arg, abc.Iterable) and not isinstance(arg, (str, bytes))
+
+@dataclass
+class ExcludeFilters:
+    ids: Set[int] = field(default_factory=set)
+    names: Set[str] = field(default_factory=set)
+    regex: Set[Pattern[str]] = field(default_factory=set)
```

Comment:
should be plural

Reply:
The `regexes` plural variant? Pluralizing it.

OFF: According to some, [the correct plural of  "regex" is "regrets"](https://twitter.com/ifosteve/status/1190348262500421634).